pkg/worker/worker.go

```diff
@@ -0,0 +1,278 @@
+	}
+
+	err = syscall.Kill(job.PID, syscall.SIGKILL)
+	if err != nil {
+		return errors.Wrapf(err, "error killing the process with pid: %d", job.PID)
+	}
+
+	job.Status = StatusStopped
+	job.StoppedAt = time.Now()
+	err = job.store()
+
+	return err
+}
+
+// Query returns information about the job.
+func (worker *Worker) Query(uuid googleUUID.UUID) (Job, error) {
+	job, err := worker.DataStorage.Retrieve(uuid)
+	if err != nil {
+		return Job{}, errors.Wrapf(err, "error retrieving job with UUID: %v", uuid)
+	}
+	if job.UUID == googleUUID.Nil {
+		return Job{}, ErrJobNotFound
+	}
+
+	return job, nil
+}
+
+// JobStatus defines a type for job statuses.
+type JobStatus string
+
+// The constants define all possible statuses of the job.
+var (
+	StatusCreated  = JobStatus("created")
+	StatusRunning  = JobStatus("running")
+	StatusError    = JobStatus("error")
+	StatusFinished = JobStatus("finished")
+	StatusStopped  = JobStatus("stopped")
+)
+
+// Job is an abstraction around Linux process that contains logs and metadata.
+type Job struct {
+	DataStorage DataStorage
+	Command     []string
+	Status      JobStatus
+	PID         int
+	UUID        googleUUID.UUID
+	CreatedAt   time.Time
+	StartedAt   time.Time
+	StoppedAt   time.Time
+	FinishedAt  time.Time
+	User        string
+	Logs        []string
+}
+
+// newJob assigns the default, required values returning new job instance.
+func newJob(dataStorage DataStorage, command []string, user string) Job {
+	return Job{
+		DataStorage: dataStorage,
+		Status:      StatusCreated,
+		Command:     command,
+		UUID:        googleUUID.New(),
+		CreatedAt:   time.Now(),
+		User:        user,
+		Logs:        []string{},
+	}
+}
+
+// start runs the the command in a new process, listens for StdIn and StdErr
+// and logs them to the DataStorage.
+// If the DataStorage is not accessible, the errors are raised up to the Worker.Start
+// method and logged to the standard output.
+// TODO: Add stopping subprocesses of the started processes.
+// TODO: Add separation of the StdIn and StdErr.
+func (job *Job) start() error {
+	cmd := exec.Command(job.Command[0], job.Command[1:]...)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		err = errors.Wrap(err, "error creating StdoutPipe for Cmd")
+		return job.logAndReturnError(err)
+	}
+	cmd.Stderr = cmd.Stdout
+
+	err = cmd.Start()
+	if err != nil {
+		err = errors.Wrap(err, "error starting process")
+		return job.logAndReturnError(err)
+	}
+
+	job.Status = StatusRunning
+	job.PID = cmd.Process.Pid
+	job.StartedAt = time.Now()
+	err = job.store()
+	if err != nil {
+		return err
+	}
+
+	buffer := make([]byte, 1000)
+	for {
+		n, err := stdout.Read(buffer)
+		if err == nil {
+			job.Logs = append(job.Logs, string(buffer[:n]))
+			err = job.store()
+			if err != nil {
+				return err
+			}
+		}
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return errors.Wrap(err, "error storing command ouput in logs")
+		}
+	}
```

Comment:
Can you think about a simpler way to do this?

You're re-implementing `io.Copy` here and you don't need to do all the copying you are doing every time you read from the reader.

Take a look at how at the implementation of [CombinedOutput](https://golang.org/src/os/exec/exec.go?s=15990:16036#L547) works. If you swap out `bytes.Buffer` for your own and pair it with a [RWMutex](https://golang.org/pkg/sync/#RWMutex) you'll get a much simpler implementation that's a lot more performant.

Reply:
I'm not sure if understand 100% correctly, so let me ask for some clarification. The key part in the CombinedOutput that you meant to show me is the direct assignment of the buffer to cmd.Stdout? This way I won't have to copy the output to the buffer as it will be copied there directly.

But I will still have to copy it later to the job.Logs anyway. If I want to present the live logs to the user, I need to update the DataStorage regularly and not only after the job is done.

I'm not sure where in the solution the RWMutex applies.